### PR TITLE
Custom environment from Stable Baselines3

### PR DESCRIPTION
**Pull request by RickFqt**
This PR adds:

- [X] Base Reward class.
- [X] Custom environment to be used when training models on Urnai with Stable Baselines3.

### CODE REVIEW

File: urnai/rewards/reward_base.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+class RewardBase(ABC):
+    """
+    Every Agent needs to own an instance of this base class in order to calculate
+    its rewards. So every time we want to create a new agent,
+    we should either use an existing RewardBase implementation or create a new one.
+    """
+
+    @abstractmethod
+    def get_reward(self, obs: list[list], default_reward: int, terminated: bool, 
+                   truncated: bool) -> int: ...
```
Comment (alvarofpp) [inline urnai/rewards/reward_base.py]:
```suggestion
    def get(
        self,
        obs: list[list],
        default_reward: int,
        terminated: bool,
        truncated: bool,
    ) -> int:
        raise NotImplementedError(...)
```

File: urnai/sc2/environments/sc2environment.py [hunk 2]
```diff
@@ -113,6 +113,6 @@ def _parse_timestep(
         terminated = any(o.player_result for o in self.env_instance._obs)
         current_steps = self.env_instance._episode_steps
         limit_steps = self.env_instance._episode_length
-        truncated = current_steps >= limit_steps
+        truncated = bool(current_steps >= limit_steps)
```
Comment (alvarofpp) [inline urnai/sc2/environments/sc2environment.py:116]: Why use `bool()` if the `>=` operator will return `bool`?

File: urnai/sc2/environments/stablebaselines3/custom_env.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                 urnai_action_space: ActionSpaceBase, reward: RewardBase, 
+                 observation_space: spaces.Space, action_space: spaces.Space):
+        super().__init__()
+
+        self._env = env
+        self._state = state
+        self._action_space = urnai_action_space
+        self._reward = reward
+        self._obs = None
+        # SB3 spaces
```
Comment (alvarofpp) [inline urnai/sc2/environments/stablebaselines3/custom_env.py]: What is `SB3`?

File: urnai/sc2/environments/stablebaselines3/custom_env.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    def reset(self, seed=None, options=None):
+        obs = self._env.reset()
+        self._obs = obs[0]
+        obs = self._state.update(self._obs)
+        info = {}
+        return obs, info
+
+    def render(self):
+        pass
```
Comment (alvarofpp) [inline urnai/sc2/environments/stablebaselines3/custom_env.py]:
```suggestion
    def render(self):
        raise NotImplementedError(...)
```

File: urnai/sc2/environments/stablebaselines3/custom_env.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    def close(self):
```
Comment (alvarofpp) [inline urnai/sc2/environments/stablebaselines3/custom_env.py]:
```suggestion
    def close(self) -> None:
```

File: urnai/sc2/environments/stablebaselines3/custom_env.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        self._env = env
+        self._state = state
+        self._action_space = urnai_action_space
+        self._reward = reward
+        self._obs = None
+        # SB3 spaces
+        self.action_space = action_space
+        self.observation_space = observation_space
+
+    def step(self, action):
```
Comment (alvarofpp) [inline urnai/sc2/environments/stablebaselines3/custom_env.py]: Enter the parameter and return types.

File: urnai/sc2/environments/stablebaselines3/custom_env.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+        obs, reward, terminated, truncated = self._env.step(action)
+
+        self._obs = obs[0]
+        obs = self._state.update(self._obs)
+        reward = self._reward.get_reward(self._obs, reward[0], terminated, truncated)
+        info = {}
+        return obs, reward, terminated, truncated, info
+
+    def reset(self, seed=None, options=None):
```
Comment (alvarofpp) [inline urnai/sc2/environments/stablebaselines3/custom_env.py]: Enter the parameter and return types.

File: urnai/sc2/environments/stablebaselines3/custom_env.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        obs = self._env.reset()
+        self._obs = obs[0]
+        obs = self._state.update(self._obs)
+        info = {}
+        return obs, info
+
+    def render(self):
```
Comment (alvarofpp) [inline urnai/sc2/environments/stablebaselines3/custom_env.py]: Enter the parameter and return types.

**Review by alvarofpp — CHANGES_REQUESTED** [review]

File: urnai/sc2/environments/sc2environment.py (hunk 2 above)
Comment (RickFqt) [reply urnai/sc2/environments/sc2environment.py:116]: It seems that this operator returns `numpy.bool`. When testing the environment with stable baselines' `check_env()` method, it raised an error requesting that `truncated` should be a python's native `bool`.

File: urnai/sc2/environments/stablebaselines3/custom_env.py (hunk 3 above)
Comment (RickFqt) [reply urnai/sc2/environments/stablebaselines3/custom_env.py]: Should be Stable Baselines3, will change in the next commit

Comment (RickFqt) [reply urnai/sc2/environments/stablebaselines3/custom_env.py]: Actually, these spaces are from `gymnasium` :D
What I wanted in the comment was to say that these variables are used internally by the library

File: urnai/sc2/environments/stablebaselines3/custom_env.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+            self, seed: int = None, options: dict = None
+        ) -> GymResetReturn:
+        obs = self._env.reset()
+        self._obs = obs[0]
+        obs = self._state.update(self._obs)
+        info = {}
+        return obs, info
+
+    def render(self, mode: str) -> None:
+        raise NotImplementedError(...)
```
Comment (alvarofpp) [inline urnai/sc2/environments/stablebaselines3/custom_env.py]: Ideally, there should be a message informing you of the error. Sorry if my previous suggestion wasn't that clear.

**Review by alvarofpp — CHANGES_REQUESTED** [review]

**Review by alvarofpp — APPROVED** [review]

**Pull request merged**